src/components/TextInput/TextInput.tsx

```diff
@@ -150,45 +148,44 @@ export const TextInput = React.forwardRef<HTMLSpanElement, TextInputProps>(funct
     };
 
     return (
-        <span
-            ref={ref}
+        <div
+            ref={ref as React.ForwardedRef<HTMLDivElement>}
             style={style}
             className={b(
                 {
                     view,
                     size,
-                    pin: view === 'clear' ? undefined : pin,
                     disabled,
                     state,
-                    'has-clear': hasClear,
+                    pin: view === 'clear' ? undefined : pin,
+                    'has-left-content': isLabelVisible,
+                    'has-right-content': isClearControlVisible && !multiline,
                     'has-scrollbar': hasVerticalScrollbar,
                 },
                 className,
             )}
             data-qa={qa}
         >
-            {isLabelVisible && (
-                <label ref={labelRef} className={b('label')} title={label} htmlFor={id}>
-                    {`${label}`}
-                </label>
-            )}
-            {multiline ? (
-                <TextAreaControl {...props} {...commonProps} controlRef={handleRef} />
-            ) : (
-                <InputControl {...props} {...commonProps} controlRef={handleRef} />
-            )}
-
+            <span className={b('content')}>
```

Comment:
Do we actually need this extra wrapper?

Reply:
Unfortunately, yes. This wrap separates input content from error message